### PR TITLE
feat(@angular-devkit/build-angular): `ng extract-i18n` support json and arb

### PR DESCRIPTION
**Pull request by wenqi73**
`@angular/localize` supports extracting as `json` file, but CLI does not support this.

### CODE REVIEW

**Comment by wenqi73** [comment]
@googlebot I fixed it.

**Comment by petebacondarwin** [comment]
If you are extracting to JSON then surely you will need to implement merging translations stored in JSON too? See https://github.com/angular/angular/blob/master/packages/localize/src/tools/src/translate/translation_files/translation_parsers/simple_json_translation_parser.ts
Also, we recently implemented ARB format, which is another form of JSON that might be worth supporting: https://github.com/angular/angular/blob/master/packages/localize/src/tools/src/extract/translation_files/arb_translation_serializer.ts and https://github.com/angular/angular/blob/master/packages/localize/src/tools/src/translate/translation_files/translation_parsers/arb_translation_parser.ts

**Comment by wenqi73** [comment]
@petebacondarwin What do you mean by "implement merging translations"? Could you tell me the details?

**Comment by petebacondarwin** [comment]
@wenqi73 - there are two main processes in i18n: extraction and merging (sometimes called translation or inlining). The extraction process involves parsing the code for `$localize` calls and rendering translation files (e.g. XLIFF) containing all the i18n messages. The merging process involves parsing the code for `$localize` calls and replacing those calls with translated messages, where the translated messages are read in from translation files (e.g. XLIFF).

When supporting a particular translation file format, we need to provide a serializer (used when extracting) and a parser (used when merging).

**Comment by wenqi73** [comment]
@petebacondarwin It already has the JSON parser in cli, https://github.com/angular/angular-cli/blob/97aa1fb7cc7595ca696fbe72fdf4982dd9815dfd/packages/angular_devkit/build_angular/src/utils/load-translations.ts#L69 so just need to add serializer? And now in my project I also use JSON files as translation files with `localize-extract`, it works fine!

**Comment by petebacondarwin** [comment]
Gadzooks! You are right! I didn't see that. Nice. OK so you are good just to add the serializer. Sorry for the noise.

**Comment by wenqi73** [comment]
@petebacondarwin I have added arb format, please have a look😄

**Review by petebacondarwin — APPROVED** [review]
LGTM

**Review by clydin — CHANGES_REQUESTED** [review]
Can you also add an E2E test file for each of the formats?
You can use the following test file as a base (and the new tests can be placed in the same directory):
https://github.com/angular/angular-cli/blob/cfb368408b532e488acaa1a5a921240f0dc7dfad/tests/legacy-cli/e2e/tests/i18n/ivy-localize-dl-xmb.ts

**Review by clydin — COMMENTED** [review]
Code changes look good.
Can you re-structure the commits so that the PR contains two feature commits: one for JSON and one for ARB (with the tests for each included in their respective commits)?

With that adjustment this should be good to be merged.
Thank you for the contribution.

**Comment by wenqi73** [comment]
@clydin Commits have been readjusted.

File: tests/legacy-cli/e2e/tests/i18n/ivy-localize-dl-json.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { executeTest } from './ivy-localize-dl-xliff2';
+import { setupI18nConfig } from './legacy';
+
+export default async function() {
+  // Setup i18n tests and config.
+  await setupI18nConfig(true, 'xmb');
```
Comment (clydin) [inline tests/legacy-cli/e2e/tests/i18n/ivy-localize-dl-json.ts:6]:
```suggestion
  await setupI18nConfig(true, 'json');
```

File: tests/legacy-cli/e2e/tests/i18n/legacy.ts [hunk 2]
```diff
@@ -226,7 +238,10 @@ export async function setupI18nConfig(useLocalize = true, format: keyof typeof f
   const translationFile = `src/locale/messages.${formats[format].ext}`;
   await expectFileToExist(translationFile);
   await expectFileToMatch(translationFile, formats[format].sourceCheck);
-  await expectFileToMatch(translationFile, `An introduction header for this sample`);
+
+  if (format !== 'json') {
```
Comment (wenqi73) [inline tests/legacy-cli/e2e/tests/i18n/legacy.ts:242]: @clydin @petebacondarwin The message description will not be extracted into JSON files.

**Review by clydin — APPROVED** [review]
LGTM.

Thanks again for the contribution.

**Pull request merged**